eth/db/chain.py

```diff
@@ -184,32 +185,36 @@ def get_block_uncles(self, uncles_hash: Hash32) -> List[BlockHeader]:
         else:
             return rlp.decode(encoded_uncles, sedes=rlp.sedes.CountableList(BlockHeader))
 
-    # TODO: This method should take a chain of headers as that's the most common use case
-    # and it'd be much faster than inserting each header individually.
-    def persist_header(self, header: BlockHeader) -> Tuple[BlockHeader, ...]:
+    def persist_header(
+            self,
+            header: Union[BlockHeader, Iterable[BlockHeader]]) -> Tuple[BlockHeader, ...]:
         """
         Returns iterable of headers newly on the canonical chain
         """
-        is_genesis = header.parent_hash == GENESIS_PARENT_HASH
-        if not is_genesis and not self.header_exists(header.parent_hash):
-            raise ParentNotFound(
-                "Cannot persist block header ({}) with unknown parent ({})".format(
-                    encode_hex(header.hash), encode_hex(header.parent_hash)))
 
-        self.db.set(
-            header.hash,
-            rlp.encode(header),
-        )
+        headers = [header] if isinstance(header, BlockHeader) else header
 
-        if is_genesis:
-            score = header.difficulty
-        else:
-            score = self.get_score(header.parent_hash) + header.difficulty
+        for header in headers:
+            is_genesis = header.parent_hash == GENESIS_PARENT_HASH
+            if not is_genesis and not self.header_exists(header.parent_hash):
+                raise ParentNotFound(
+                    "Cannot persist block header ({}) with unknown parent ({})".format(
+                        encode_hex(header.hash), encode_hex(header.parent_hash)))
 
-        self.db.set(
-            SchemaV1.make_block_hash_to_score_lookup_key(header.hash),
-            rlp.encode(score, sedes=rlp.sedes.big_endian_int),
-        )
+            self.db.set(
+                header.hash,
+                rlp.encode(header),
+            )
+
+            if is_genesis:
+                score = header.difficulty
+            else:
+                score = self.get_score(header.parent_hash) + header.difficulty
```

Comment:
We can probably avoid this `get_score()` call now, as we're processing a chain of headers so we just need to get the score of the first item and then increment that with the difficulty of each subsequent item

Reply:
We will need to do one of the following here:

- pre-validate that the iterable of headers is a contiguous chain.
- handle non-contiguous chains but still retain this optimizations for when it is a chain, i.e. detect when we are in a sequential list of headers and just increment the score, as well as detecting when we are non-sequential and lookup the score in this manner.